### PR TITLE
EG-548 end point to save page review

### PR DESCRIPTION
**Pull request by LesWilson**

### CODE REVIEW

File: app/repositories/ApprovalProcessReviewRepository.scala [hunk 1]
```diff
@@ -89,4 +91,31 @@ class ApprovalProcessReviewRepositoryImpl @Inject() (implicit mongoComponent: Re
...
+
+    val selector = Json.obj("ocelotId" -> id, "version" -> version, "pages.pageUrl" -> pageUrl)
+    val modifier =
+      Json.obj(
+        "$set" -> Json.obj(
+          "pages.$.result" -> reviewInfo.result,
+          "pages.$.status" -> reviewInfo.status,
+          "pages.$.comment" -> reviewInfo.comment,
+          "pages.$.updateUser" -> reviewInfo.updateUser,
+          "pages.$.updateDate" -> Json.obj("$date" -> LocalDateTime.now().toInstant(ZoneOffset.UTC).getEpochSecond * 1000)
```
Comment (cmrhmrc) [inline app/repositories/ApprovalProcessReviewRepository.scala:104]: LocalDateTime.now().atZone(ZoneId.of("UTC")).toInstant().toEpochMilli()

File: app/controllers/ProcessReviewController.scala [hunk 2]
```diff
@@ -66,4 +66,19 @@ class ProcessReviewController @Inject() (reviewService: ReviewService, cc: Contr
     }
   }
 
+  def approval2iReviewPageComplete(id: String, pageUrl: String): Action[JsValue] = Action.async(parse.json) { request =>
+    def save(reviewInfo: ApprovalProcessPageReview): Future[Result] = {
```
Comment (cmrhmrc) [inline app/controllers/ProcessReviewController.scala:70]: redundant {}

File: app/services/PublishedService.scala [hunk 3]
```diff
@@ -53,14 +53,18 @@ class PublishedService @Inject() (repository: PublishedRepository) {
 
     def saveProcess: Future[RequestOutcome[String]] = {
```
Comment (cmrhmrc) [inline app/services/PublishedService.scala:54]: redundant {}

File: app/services/ReviewService.scala [hunk 4]
```diff
@@ -64,31 +64,61 @@ class ReviewService @Inject() (publishedService: PublishedService, repository: A
   def changeStatus(id: String, currentStatus: String, info: ApprovalProcessStatusChange): Future[RequestOutcome[Unit]] = {
 
     def getContentToUpdate: Future[Option[ApprovalProcess]] = repository.getById(id) map {
-        case Right(process) => if (process.meta.status == currentStatus) Some(process) else None
-        case _ => None
-      }
+      case Right(process) =>
+        if (process.meta.status == currentStatus) {
```
Comment (cmrhmrc) [inline app/services/ReviewService.scala:68]: clearer as 
case Right(process) if process.meta.status == currentStatus =>
case Right(process) =>
case Left(err) =>

File: app/services/ReviewService.scala [hunk 5]
```diff
@@ -64,31 +64,61 @@ class ReviewService @Inject() (publishedService: PublishedService, repository: A
...
+          Some (process)
+        }
+        else {
+          logger.warn(s"Invalid Process Status Change requested for process $id: " +
+            s"Expected Status: '$currentStatus' Status Found: '${process.meta.status}' Desired Status: ${info.status} ")
+          None
+        }
+      case _ =>
```
Comment (cmrhmrc) [inline app/services/ReviewService.scala:76]: Could be a db err from the getById() call

File: app/repositories/ApprovalProcessReviewRepository.scala (hunk 1 above)
Comment (LesWilson) [reply app/repositories/ApprovalProcessReviewRepository.scala:104]: I knew there must be a way to get millis, I didn't look hard enough :) thanks

File: app/controllers/ProcessReviewController.scala (hunk 2 above)
Comment (LesWilson) [reply app/controllers/ProcessReviewController.scala:70]: still got my java head on ....... 20+ years of curly braces difficult to drop!!

File: app/services/ReviewService.scala (hunk 4 above)
Comment (LesWilson) [reply app/services/ReviewService.scala:68]: thanks, it didn't seem the right way to structure it

File: app/services/ReviewService.scala [hunk 6]
```diff
@@ -64,31 +64,61 @@ class ReviewService @Inject() (publishedService: PublishedService, repository: A
...
+        publishedService.save(id, approvalProcess.process) map {
+          case Right(_) => Right(())
+          case Left(errors) =>
+            logger.error(s"Failed to publish $id - $errors")
+            Left(errors)
+        }
+      case _ => Future.successful(Right(()))
+    }
 
     getContentToUpdate flatMap {
```
Comment (cmrhmrc) [inline app/services/ReviewService.scala:92]: You've converted to Option return and lost potential errors which may have occured in the RequestOutcome calls, sticking to RequestOutcome throughout will ensure fast fail and errors returned

File: app/services/ReviewService.scala [hunk 7]
```diff
@@ -64,31 +64,61 @@ class ReviewService @Inject() (publishedService: PublishedService, repository: A
...
+            Future.successful(error)
           case _ => publishIfRequired(approvalProcess)
         }
       case None =>
-        logger.warn(s"Invalid process id submitted to method changeStatus. The requested id was $id")
         Future.successful(Left(Errors(NotFoundError)))
     }
   }
 
+  def approval2iReviewPageComplete(id: String, pageUrl: String, reviewInfo: ApprovalProcessPageReview): Future[RequestOutcome[Unit]] = {
```
Comment (cmrhmrc) [inline app/services/ReviewService.scala:107]: redundant {}

File: app/repositories/ApprovalProcessReviewRepository.scala (hunk 1 above)
Comment (LesWilson) [reply app/repositories/ApprovalProcessReviewRepository.scala:104]: Done - in next commit

File: app/controllers/ProcessReviewController.scala (hunk 2 above)
Comment (LesWilson) [reply app/controllers/ProcessReviewController.scala:70]: Done, in next commit

File: app/services/PublishedService.scala (hunk 3 above)
Comment (LesWilson) [reply app/services/PublishedService.scala:54]: Done, in next commit

File: app/services/ReviewService.scala (hunk 4 above)
Comment (LesWilson) [reply app/services/ReviewService.scala:68]: Done, in next commit

File: app/services/ReviewService.scala (hunk 5 above)
Comment (LesWilson) [reply app/services/ReviewService.scala:76]: restructured to handle errors

File: app/services/ReviewService.scala (hunk 6 above)
Comment (LesWilson) [reply app/services/ReviewService.scala:92]: updated

File: app/services/ReviewService.scala (hunk 7 above)
Comment (LesWilson) [reply app/services/ReviewService.scala:107]: done

**Pull request merged**